### DISCUSSION TITLE
CoreWCF Service Running in Azure Container App - Hitting Endpoints Returns HTTP 404

### DISCUSSION THREAD

**Discussion by essenbee2** [CoreWCF/CoreWCF#1515] [Q&A] [answered]
I have a .NET 8.0 CoreWCF Service that exposes WCF and REST endpoints for a set of service methods. This runs fine when hosted by IIS using the .NET 8.0 hosting package. However, we are looking to migrate to Azure Container Apps using .NET Aspire and `azd` to deploy. The deployment works great, and I can see my container running. The output on the console is as follows:

```
info: CoreWCF.Channels.ServiceModelHttpMiddleware[0]
      Mapping CoreWCF branch app for path /ContactServiceCore.svc
info: CoreWCF.Channels.ServiceModelHttpMiddleware[0]
      Mapping CoreWCF branch app for path /api
info: CoreWCF.Channels.MetadataMiddleware[0]
      Configuring metadata to /ContactServiceCore.svc
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://[::]:8080
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Development
info: Microsoft.Hosting.Lifetime[0]
      Content root path: /app
```

From this, it would appear that my endpoints are `/ContactServiceCore.svc` and `/api`, but trying to view the WSDL on the first endpoint or exercise service methods on the second just return 404. However, I can view the Swagger UI and get a response from the /health endpoint set up by .NET Aspire Service Defaults.

The container is running fine at https://contact-service.***********.*******.azurecontainerapps.io

The Ingress for the container looks like this:

<img width="580" alt="image" src="https://github.com/user-attachments/assets/a0566438-6347-40e1-9f58-6195938bdf10" />

and my service endpoints are being created in `Program.cs` like this:

```
// Configure the binding to have a very large buffer
var veryLargeBufferBinding = useHttps
    ? new BasicHttpBinding(BasicHttpSecurityMode.Transport)
    : new BasicHttpBinding();

var apiBinding = useHttps 
    ? new WebHttpBinding(WebHttpSecurityMode.Transport)
    : new WebHttpBinding();


app.UseServiceModel(serviceBuilder =>
{
    serviceBuilder.AddService<ContactService.ContactService>();
    serviceBuilder.AddServiceEndpoint<ContactService.ContactService, IContactService>(veryLargeBufferBinding, "/ContactServiceCore.svc");

    // REST endpoint
    serviceBuilder.AddServiceWebEndpoint<ContactService.ContactService, IContactService>(apiBinding, "api", behavior =>
    {
        behavior.HelpEnabled = app.Environment.IsDevelopment();
    });

    var serviceMetadataBehavior = app.Services.GetRequiredService<ServiceMetadataBehavior>();

    if (useHttps)
    {
        serviceMetadataBehavior.HttpsGetEnabled = true;
    }
    else
    {
        serviceMetadataBehavior.HttpGetEnabled = true;
    }
});
```

Note that I have `useHttps = false` because if I set it to `true` the container will not start and throws the exception:

`Unhandled exception. System.InvalidOperationException: Could not find a base address that matches scheme https for the endpoint with binding BasicHttpBinding. Registered base address schemes are [http].
   at CoreWCF.ServiceHostBase.MakeAbsoluteUri(Uri relativeOrAbsoluteUri, Binding binding, UriSchemeKeyedCollection baseAddresses)`

So:

`https://contact-service.***********.*******.azurecontainerapps.io/ContactServiceCore.svc?wsdl` returns 404
`https://contact-service.***********.*******.azurecontainerapps.io/api/GetContactByEmailAddress/s*******@*****.org` returns 404

I have tried endless variations of these URLs to no avail. However:

`https://contact-service.***********.*******.azurecontainerapps.io/Swagger.Index.html` displays the Swagger UI
`https://contact-service.***********.*******.azurecontainerapps.io/health` displays the text `Healthy`

I am at a loss how to resolve this and would gratefully accept any help. Please note that I am not an expert in containers or Azure Container Apps. I was hoping that .NET Aspire and `azd` would "just work" and for the most part they have been great -0 I'm just stuck with a running container that returns 404 when I try to use the CoreWCF service methods!

**Comment by alfonsovgs** [comment, marked as answer]
Hi @essenbee2,

Is your project running locally? I had that error; this is my settings.


```
app.UseServiceModel(cfg =>
{
    string host = builder.Configuration.GetValue<string>("Host");
    cfg.AddService<CustomerService>(opts =>
            {
                opts.DebugBehavior.IncludeExceptionDetailInFaults = true;
                opts.DebugBehavior.HttpsHelpPageEnabled = true;
                opts.BaseAddresses.Add(new Uri($"http://{host}/CustomerService.svc"));
                opts.BaseAddresses.Add(new Uri($"https://{host}/CustomerService.svc"));
            });

    cfg.AddServiceEndpoint<CustomerService, ICustomerService>(new BasicHttpBinding(BasicHttpSecurityMode.Transport), string.Empty);
});

var serviceMetadataBehavior = app.Services.GetRequiredService<ServiceMetadataBehavior>();
serviceMetadataBehavior.HttpGetEnabled = serviceMetadataBehavior.HttpsGetEnabled = true;

```

appsettings.json
`"Host": "localhost",`

**Reply by essenbee2** [reply]
Hi @alfonsovgs. Yes, it runs fine locally and when hosted in IIS. The problem with having a Host env var is that I don't know what the URL of the container app is until after I deploy it via `azd`. However, I do know it now the `azd provision` step is done so I will give your suggestion a try. Thanks for responding!

**Reply by essenbee2** [reply]
@alfonsovgs or are you saying that, in the container, the Host will in fact be `localhost`?

**Reply by alfonsovgs** [reply]
> @alfonsovgs or are you saying that, in the container, the Host will in fact be `localhost`?

Of course, if I remember correctly, the value should always be localhost. In the case of IIS, if you have sites, they should go as localhost/YOUR_SITE_NAME.

**Reply by essenbee2** [reply]
@alfonsovgs Thank you! I will give that a try tomorrow and report back here. Appreciate your input.

**Reply by mconnew** [reply]
There's an annoying thing that Azure sometimes does with some of their newer hosting models. They place a reverse proxy server listening for https requests in front of your service and forward the requests to your app. Kestrel actually listens on an http address as the requests are forwarded over http, but they insert some middleware which fixes some things up and presents the request as though it arrived over https. This puts CoreWCF in a weird situation as it requires an https listening address from Kestrel, but that says there's only an http address listening. So you think okay, I'll switch to using an HTTP binding (not HTTPS), but then the requests arriving say they are using HTTPS so they can't be dispatched by CoreWCF as that doesn't match.  
If this is what you're facing, configure everything as though you are using https, then via an overload of AddService\<TService\> which takes a delegate which passes a ServiceOptions class, provide a delegate which clears out ServiceOptions.BaseAddresses (there's a Clear() method) and add a dummy https base address. CoreWCF doesn't care about the port number, just wants the scheme to match. This will then tell CoreWCF to ignore what Kestrel told it and you do actually have an HTTPS address that you are going to be receiving requests from.

**Reply by mconnew** [reply]
To clarify a few things, WCF and CoreWCF by default don't care about the hostname or the port number, only the scheme (http vs https). There is a HostnameComparisonMode property you can set to make it care, but I'm not sure if I even implemented anything behind that. The main purpose of HostnameComparisonMode is to enable you to have 2 different services at the same url path for 2 different hostnames. That's not what you're doing here, so just use https://localhost/ and/or http://localhost/ and don't worry about the hostname or port number. If you care about rejecting requests based on the hostname, I would suggest doing it using some inline custom middleware instead of having CoreWCF handle it.

**Reply by essenbee2** [reply]
Thank you so much @alfonsovgs and @mconnew for the assistance. I can now see the WSDL for the WCF service. Just need to figure out how to get the REST api working now, as calling it gives a 400 Bad Request :-(

**Reply by essenbee2** [reply]
Got the REST API working now :-) It was the situation where you need to order the "add service endpoints" by the longest path first.

**Reply by miguelEsteban** [reply]
I need to hardcoded schema in a middleware too.

```
app.Use(async (context, next) =>
{
    IHttpRequestFeature? httpRequestFeature = context.Features.Get<IHttpRequestFeature>();
    if (httpRequestFeature != null)
    {
        httpRequestFeature.Scheme = "https";
    }
    await next();
});
```
